transparency-in-coverage/python/processors/helper_suggestions/state_management_flattener.py

```diff
@@ -0,0 +1,320 @@
+import csv
+import os
+from io import FileIO
+
+import ijson
+
+from processors.helper_suggestions.hash_dict import hashdict
+from processors.helpers import MRFOpen, Row
+from processors.schema import SCHEMA
+
+
+class StateManagedFlattener:
+    """
+    Bundled methods for handling the flattening
+    of streamed JSON.
+    """
+
+    def __init__(
+        self, code_set: list[tuple[str, str]] = None, npi_set: set[int] | None = None
+    ):
+        """ """
+        self.npi_set = {} if npi_set is None else npi_set
+        self.code_set = {} if code_set is None else code_set
+
+        self.local_provider_references = None
+        self.remote_provider_references = []
+
+        self.provider_reference_map = None
+        self.root_written = False
+        self.in_network_item = None
+
+    @classmethod
+    def factory(
+        cls, f: FileIO, code_set: list[tuple[str, str]], npi_set: set[int] | None
+    ) -> "StateManagedFlattener":
+        flattener = cls(code_set, npi_set)
+        flattener._init_parser(f)
+
+        # Build (but don't write) the root data
+        flattener._build_root()
+
+        # Jump (fast-forward) to provider references
+        # Build (but don't write) the provider references
+        flattener.ffwd(("", "map_key", "provider_references"))
+        flattener._build_provider_references()
+        flattener._build_remote_provider_references()
+        flattener._make_provider_ref_map()
+        return flattener
+
+    def ffwd(self, to_row):
+        """Jump to later in the parsing stream"""
+        while self.current_row != to_row:
+            self.current_row = next(self.parser)
+
+    def _init_parser(self, f):
+        self.parser = ijson.parse(f, use_float=True)
+
+    def _build_root(self):
```

Comment:
All methods that therefore are not used outside this class are marked as private. This also goes a long way towards showing another user what are the "main" methods of the class that they should know about, and which are internal details they don't necessarily need to know.

Reply:
This is legit as fuck